tiled/_tests/conftest.py

```diff
@@ -105,29 +107,74 @@ def poll_enumerate():
 
     Note that startup() and shutdown() are not called, and must be run
     either manually (as in the fixture 'a') or via the app (as in the fixture 'client').
     """
-    if request.param == "sqlite":
-        adapter = in_memory(writable_storage=str(tmpdir))
+    if not TILED_TEST_POSTGRESQL_URI:
+        raise pytest.skip("No TILED_TEST_POSTGRESQL_URI configured")
+    # Create temporary database.
+    async with temp_postgres(TILED_TEST_POSTGRESQL_URI) as uri_with_database_name:
+        # Build an adapter on it, and initialize the database.
+        adapter = from_uri(
+            uri_with_database_name,
+            writable_storage=str(tmpdir),
+            init_if_not_exists=True,
+        )
         yield adapter
-    elif request.param == "postgresql":
-        if not TILED_TEST_POSTGRESQL_URI:
-            raise pytest.skip("No TILED_TEST_POSTGRESQL_URI configured")
-        # Create temporary database.
-        async with temp_postgres(TILED_TEST_POSTGRESQL_URI) as uri_with_database_name:
-            # Build an adapter on it, and initialize the database.
-            adapter = from_uri(
-                uri_with_database_name,
-                writable_storage=str(tmpdir),
-                init_if_not_exists=True,
-            )
-            yield adapter
-            await adapter.shutdown()
-    else:
-        assert False
+
+
+@pytest_asyncio.fixture
+async def postgresql_with_example_data_adapter(request, tmpdir):
+    """
+    Adapter instance
+
+    Note that startup() and shutdown() are not called, and must be run
+    either manually (as in the fixture 'a') or via the app (as in the fixture 'client').
+    """
+    if not TILED_TEST_POSTGRESQL_URI:
+        raise pytest.skip("No TILED_TEST_POSTGRESQL_URI configured")
+    DATABASE_NAME = "example_data"
+    uri = TILED_TEST_POSTGRESQL_URI
+    if uri.endswith("/"):
+        uri = uri[:-1]
+    uri_with_database_name = f"{uri}/{DATABASE_NAME}"
+    engine = create_async_engine(uri_with_database_name)
+    try:
+        async with engine.connect():
+            pass
+    except asyncpg.exceptions.InvalidCatalogNameError:
+        raise pytest.skip(
+            f"PostgreSQL instance contains no database named {DATABASE_NAME!r}"
+        )
+    adapter = from_uri(
+        uri_with_database_name,
+        writable_storage=str(tmpdir),
+    )
+    yield adapter
+
+
+@pytest_asyncio.fixture
+async def sqlite_adapter(request, tmpdir):
+    """
+    Adapter instance
+
+    Note that startup() and shutdown() are not called, and must be run
+    either manually (as in the fixture 'a') or via the app (as in the fixture 'client').
+    """
+    yield in_memory(writable_storage=str(tmpdir))
+
+
+@pytest.fixture(params=["sqlite_adapter", "postgresql_adapter"])
+def adapter(request):
+    """
+    Adapter instance
+
+    Note that startup() and shutdown() are not called, and must be run
+    either manually (as in the fixture 'a') or via the app (as in the fixture 'client').
```

Comment:
For ease of use, it seems like it would be convenient to keep `startup()` and `shutdown()` in the fixtures. Is it straightforward to add a guard to `startup()` that checks whether it has already been called or whether an app is running?

Reply:
I do dislike this structure and would welcome suggestions to improve it.

The problem is that startup must be called on the same thread where the application will run. If this adapter is going to be used by the `TestClient, via `Context.from_app(build_app(adapter))`, a background thread is created at that point and `startup()` needs to be run on that thread.

Reply:
I see.  In that case, I think the solution using the fixture `a` is probably already optimal.

---

One could of course add fixture `b` (or equivalent) for `test_metadata_index_is_used(b)` -- if you expect additional tests that would make use of `postgresql_with_example_data_adapter`:

```python3
@pytest_asyncio.fixture
async def b(postgresql_with_example_data_adapter):
    "Raw adapter, not to be used within an app becaues it is manually started and stopped."
    adapter = postgresql_with_example_data_adapter
    await adapter.startup()
    yield adapter
    await adapter.shutdown()
```


However, a DRYer and more composable version might look like this...

```python3
@pytest.mark.parametrize("a", ["postgresql_with_example_data_adapter"], indirect=True)
@pytest.mark.asyncio
async def test_metadata_index_is_used(a):
    # a, a.startup(), a.shutdown() are no longer needed
    ...
```

Marking the parameter as `indirect` will override the argument passed to a parameterized fixture.  See ["Indirect parametrization"](https://docs.pytest.org/en/7.1.x/example/parametrize.html#indirect-parametrization) | [`parametrize`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.Metafunc.parametrize) | and especially [this informative example](https://stackoverflow.com/a/77061010).